db/compaction/compaction_picker.cc

```diff
@@ -656,7 +671,9 @@ Compaction* CompactionPicker::CompactRange(
       inputs[level - start_level].level = level;
       auto& files = inputs[level - start_level].files;
       for (FileMetaData* f : vstorage->LevelFiles(level)) {
-        files.push_back(f);
+        if (output_level > 0 || f->fd.largest_seqno <= earliest_mem_seqno) {
```

Comment:
oh okay - that seems like a contradiction between CompactRange() contract (i.e, "a key should be compacted .... ") and the "skip file" approach we are pursuing here .... Have we discussed the solution to this before? 

Maybe leave as TODO for CompactionPicker ::CompactRange() for now and fix it by not allowing such "L0 file overlapping memtable in seqno" exists as you mentioned in the task about next step for universal compaction?

Reply:
No solution has been discussed before, but a necessary condition for this bug ("If the memtable were not flushed, ...") is not present in universal CompactRange(), because it always does the flush. My original point was just that this is a big assumption that might change or might not even be true if I misread the code.